### PR TITLE
Grafana UI: Add experimental InteractiveTable component

### LINKED ISSUE #58942: Grafana-ui: Add InteractiveTable to component library
We need to move the table used for the correlations settings page to `@grafana/ui`

Basic requirements are described more in detail in https://docs.google.com/document/d/1D9jnhYj1ZtxNHYhDJgDv4i--43qrBjh5KJX7wIdGf1k/edit#heading=h.k7kv6iubst6n

### PR DESCRIPTION
**Pull request by Elfo404**
Adds an `InteractiveTable` component based on the requirements described in https://docs.google.com/document/d/1D9jnhYj1ZtxNHYhDJgDv4i--43qrBjh5KJX7wIdGf1k/edit#heading=h.k7kv6iubst6n

The implementation is missing some of the features described in the doc, the idea is to add them gradually and move forward with this as discussed offline with @radiorental @staton-hyse11 and @JoaoSilvaGrafana 

Fixes #58942

Notes: 
- The component is marked as `alpha` and lives inside an `Experimental` folder. 
- utility types in `public/app/core/utils/types.ts` are moved to grafana/data to ease reusability, files using it are also updated
- columns sorting is based on an example implementation described in https://www.w3.org/WAI/ARIA/apg/example-index/table/sortable-table which should make the feature accessible

### CODE REVIEW

File: packages/grafana-ui/src/components/DataTable/types.ts [hunk 1]
```diff
@@ -0,0 +1,34 @@
...
+   */
+  header?: string;
+  /**
+   * Column sort type. If `undefined` the column will not be sortable.
+   * */
+  sortType?: DefaultSortTypes | SortByFn<TableData>;
+  /**
+   * If `true` the column will shrink to its minimum width.
+   */
+  shrink?: boolean;
```
Comment (JoaoSilvaGrafana) [inline packages/grafana-ui/src/components/DataTable/types.ts:26]: This naming here isn't very clear to me, I thought it was related to flex-shrink. The behavior is actually more of a `grow: false`, wdyt about calling it `canGrow` and reverse the `boolean` meaning?

**Review by JoaoSilvaGrafana — COMMENTED** [review]
Nice work on extracting all this and adding docs! 🎉 
Left some comments and have one other comment in that I find the multi-sorting generally confusing... At first I thought it mattered the order in which I clicked the columns to sort the row, but now I see that if there are more than 2 columns sorted it just takes the sorting of the first column? So I don't think that is very useful. Do we have a requirement for that to be possible?

File: packages/grafana-ui/src/components/DataTable/types.ts [hunk 2]
```diff
@@ -0,0 +1,34 @@
...
+  /**
+   * If the provided function returns `false` the column will be hidden.
+   */
+  visible?: (data: TableData[]) => boolean;
+}
+
+type Truthy<T> = T extends false | '' | 0 | null | undefined ? never : T;
+export const isTruthy = <T>(value: T): value is Truthy<T> => Boolean(value);
```
[the remark this answers is not in the document]
Comment (Elfo404) [reply packages/grafana-ui/src/components/DataTable/types.ts:34]: forgot to mention this 🤦 i copied them over from core because we can't import stuff from core in `@grafana/ui`. i was actually thinking of moving them to `@grafana/data` (given both ui and core depends on it) and import it from there. wdyt?

**Comment by Elfo404** [comment]
@JoaoSilvaGrafana agree with the multisort being confusing (and not working properly). i disabled it, we can get back to it if we actually need it.

File: packages/grafana-ui/src/components/DataTable/types.ts [hunk 3]
```diff
@@ -0,0 +1,34 @@
+import { ReactNode } from 'react';
+import { CellProps, DefaultSortTypes, IdType, SortByFn } from 'react-table';
+
+export interface Column<TableData extends object> {
+  /**
+   * ID of the column.
+   * Set this to the matching object key of your data or `undefined` if the column doesn't have any associated data with it.
+   * This must be unique among all other columns.
+   */
+  id?: IdType<TableData>;
```
[the remark this answers is not in the document]
Comment (Elfo404) [reply packages/grafana-ui/src/components/DataTable/types.ts:10]: technically it is used to match the column with the relevant piece of data. setting it to undefined will cause the underlying data to not match with any column, but i agree is confusing. may it's better to just make it required? if the key does not exist in the underlying data there won't be any issue, the only reason i made it option was to avoid users to have to define an id when they didn't need it, but again, maybe making it required would simplify reasoning about it. wdyt?

File: packages/grafana-ui/src/components/DataTable/types.ts (hunk 1 above)
Comment (Elfo404) [reply packages/grafana-ui/src/components/DataTable/types.ts:26]: i personally think optional boolean props/options are alway easier to reason about when their default is `false` and by setting them to `true` you trigger a behaviour that wouldn't otherwise be triggered.

in this case, being it an optional boolean property i would expect not passing it to behave the same as setting it to false.

while i agree that `shrink` may not be the best name, i always have an hard time having to "inverse" how i think about props/options when some of them reverse this principle.

what about `disableGrow`? or maybe `narrow` (which i kinda like as it sounds more declarative)?

File: packages/grafana-ui/src/components/DataTable/types.ts (hunk 3 above)
[the remark this answers is not in the document]
Comment (JoaoSilvaGrafana) [reply packages/grafana-ui/src/components/DataTable/types.ts:10]: Yeah I don't think it's that much worse to force every column to have an id and it might save us from other issues in the future so making it required sounds like a good idea!

File: packages/grafana-ui/src/components/DataTable/types.ts (hunk 1 above)
Comment (JoaoSilvaGrafana) [reply packages/grafana-ui/src/components/DataTable/types.ts:26]: > i personally think optional boolean props/options are alway easier to reason about when their default is false and by setting them to true you trigger a behaviour that wouldn't otherwise be triggered.
Yep you're right those do make more sense!

I personally prefer `disableGrow` to `narrow`, because with `narrow` it makes it sound like the column will be really tiny, when really it can be gigantic depending on the title 😂  But maybe it's just a personal preference.

Comment (Elfo404) [reply packages/grafana-ui/src/components/DataTable/types.ts:26]: good point. i agree 👍

File: packages/grafana-ui/src/components/DataTable/DataTable.story.tsx [hunk 4]
```diff
@@ -0,0 +1,137 @@
...
+      exclude: EXCLUDED_PROPS,
+    },
+  },
+  args: {},
+  argTypes: {},
+};
+
+interface TableData {
+  header1: string;
+  header2?: number;
```
[the remark this answers is not in the document]
Comment (Elfo404) [reply packages/grafana-ui/src/components/DataTable/DataTable.story.tsx:30]: good catch, fixed it in 3e71e9acfc

File: packages/grafana-ui/src/components/DataTable/types.ts (hunk 3 above)
[the remark this answers is not in the document]
Comment (Elfo404) [reply packages/grafana-ui/src/components/DataTable/types.ts]: changed in 2a8aceaed5

File: packages/grafana-ui/src/components/DataTable/DataTable.mdx [hunk 5]
```diff
@@ -0,0 +1,115 @@
...
+
+<Meta title="MDX|DataTable" component={DataTable} />
+
+# DataTable
+
+<Badge text="Alpha" icon="rocket" color="blue" tooltip="This component is still experimental." />
+
+The DataTable is used to display and select data efficiently.
+It allows for the display and modification of detailed information.
+With additional functionality it allows for batch editing, as needed by your feature's users.
```
[the remark this answers is not in the document]
Comment (Elfo404) [reply packages/grafana-ui/src/components/DataTable/DataTable.mdx:14]: added some more docs in f65efa953f

File: packages/grafana-ui/src/components/DataTable/types.ts [hunk 6]
```diff
@@ -0,0 +1,31 @@
+import { ReactNode } from 'react';
+import { CellProps, DefaultSortTypes, IdType, SortByFn } from 'react-table';
+
+export interface Column<TableData extends object> {
+  /**
+   * ID of the column.
+   * Set this to the matching object key of your data or `undefined` if the column doesn't have any associated data with it.
```
Comment (JoaoSilvaGrafana) [inline packages/grafana-ui/src/components/DataTable/types.ts:7]: Should remove the mention of `undefined` from here

**Review by JoaoSilvaGrafana — COMMENTED** [review]
It looks good minus that one small comment change, great job on this!

File: packages/grafana-ui/src/components/DataTable/types.ts (hunk 6 above)
Comment (Elfo404) [reply packages/grafana-ui/src/components/DataTable/types.ts:7]: 🤦 of course i forgot this

File: packages/grafana-ui/src/components/InteractiveTable/InteractiveTable.tsx [hunk 7]
```diff
@@ -39,13 +33,33 @@ const getStyles = (theme: GrafanaTheme2) => ({
...
-    top: ${theme.spacing(1)};
+  header: css`
+    position: relative;
+    &,
+    & > button {
+      position: relative;
+      white-space: nowrap;
+      padding: ${theme.spacing(1)};
+    }
+    & > button {
```
Comment (JoaoSilvaGrafana) [inline packages/grafana-ui/src/components/InteractiveTable/InteractiveTable.tsx:44]: Is there a reason for the two different `& > button`? Specificity things?

**Review by JoaoSilvaGrafana — COMMENTED** [review]
Got a general question about the accessibility of the table.
- It doesn't seem like the content of the table is read in a screen reader (VoiceOver at least)
- Same when a row is expanded, the expanded content should be read
- Cannot tab into a column header that can be sorted
- When tabbing into a row expander, do we think it should read the row number that it is expanding? Otherwise it is hard to follow which one the user is expanding

I don't fully know all the ways tables should be accessible, those were just some things I noted when trying it with keyboard only or screen reader. wdyt?

File: packages/grafana-ui/src/components/InteractiveTable/InteractiveTable.tsx (hunk 7 above)
Comment (Elfo404) [reply packages/grafana-ui/src/components/InteractiveTable/InteractiveTable.tsx:43]: the first set of rules also applies to the `th`: `&, & > button`, while the second only to the button. there was an extra `position: relative;` at line 37 tho that i removed in 8aa4df9696

**Comment by Elfo404** [comment]
@JoaoSilvaGrafana thanks for looking at this again!

> Cannot tab into a column header that can be sorted

i can:
https://user-images.githubusercontent.com/1170767/202770802-737a893d-c1da-4b93-bfd2-83c60ac8c0cc.mov

are you sure the example you are testing it on has sortable columns? only `basic` does.

regarding
> It doesn't seem like the content of the table is read in a screen reader (VoiceOver at least)
and in general when reading content, non interactive elements are only read by voiceover when it gets told to;

to read the content of a column (and its sorting state) you can press `option+control+c`, the first time it will only read the content of the header (button) the second time it will read the column content, you can check a demo here:

https://user-images.githubusercontent.com/1170767/202774469-8fac6cff-921c-4e5c-8c46-f8d15869cd04.mov


in general you can refer to https://dequeuniversity.com/screenreaders/voiceover-keyboard-shortcuts and in https://www.apple.com/voiceover/info/guide/_1133.html#:~:text=Here%20are%20ways%20to%20navigate,header%2C%20press%20VO%2DC. you can find a section explaining how to read table contents. lmk if you want to look at this together!


similar thing about
>When tabbing into a row expander, do we think it should read the row number that it is expanding? Otherwise it is hard to follow which one the user is expanding

voiceover and in general other screenereaders should as well) already tells this information.
you can use `option+control+c` to describe the content of the element, and `option+control+r` to read the row.


does this answer your question? (of course all of the above is based on my (very limited) experience with voiceover, i'm far from being an expert, i may be completely wrong)

**Comment by Elfo404** [comment]
@JoaoSilvaGrafana good question! in general you can navigate non-interactive content in many ways. for tables you can use `option+control+[arrow]`. I slightly changed one of the examples to include some text before the table just to show how i can move to and navigate the table **without** tabbing into it (i don't have keycastr on, but i promise i didn't press tab 😛 ). as you can see i'm moving across cells in the same row (where the header is announced correctly before the content of the cell, i.e. "Data Source, Prometheus"), expanding rows, navigating to the expanded content or even navigating across cells in a column (`control+option=arrow up/down`)

https://user-images.githubusercontent.com/1170767/203033220-bcd4fdce-925d-48a6-b81e-a2d873f6aa6c.mov

**Comment by JoaoSilvaGrafana** [comment]
This is very cool ❤️

**Review by JoaoSilvaGrafana — APPROVED** [review]
👏 👏
